Triggered MC Info

StevenGreen1 opened the pull request:
Access the MCTruth block produced by the generator to determine, which particles (non-neutrino) are primary (i.e. triggered beam particles for ProtoDUNE).

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraHelper.cxx:
```diff
@@ -735,6 +735,39 @@ void LArPandoraHelper::CollectMCParticles(const art::Event &evt, const std::stri
+    for (int i = 0; i < mcTruth->NParticles(); ++i)
+    {
+        const simb::MCParticle particle = mcTruth->GetParticle(i); 
+        particleVector.push_back(particle);
```
Can we avoid making a temporary copy `particle` here?

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraHelper.h:
```diff
@@ -354,6 +355,16 @@ class LArPandoraHelper
     static void CollectMCParticles(const art::Event &evt, const std::string &label, MCParticleVector &particleVector);
 
+    // ATTN:  This function is needed as accessing generator (opposed to Geant4) level MCParticles requires use of MCTruth block.
+    /**
```
Should probably move the ATTN to part of the doxygen brief.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraInput.cxx:
```diff
@@ -469,6 +469,10 @@ void LArPandoraInput::CreatePandoraMCParticles(const Settings &settings, const M
     int particleCounter(0);
 
+    // Find Primary Generator Particles
+    std::map<const simb::MCParticle, bool> primaryGeneratorMCParticleMap;
```
Do you know how the ordering works for a map keyed on a `simb::MCParticle` - does it have an `operator<`?

StevenGreen1 replied:
Hi John,

Yes, the return value for the operator< is calculated based on the trackID of the simb::MCParticle.  At generator level all trackIDs are negative numbers that get progressivelt more negative i.e. triggered beam particle is -1 and background beam particles are -2, -3, -4 etc. Geant4 gives all MCParticles a positive track ID.

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraInput.cxx:
```diff
@@ -572,6 +580,41 @@ void LArPandoraInput::CreatePandoraMCParticles(const Settings &settings, const M
 
+void LArPandoraInput::FindPrimaryParticles(const RawMCParticleVector &mcParticleVector, std::map<const simb::MCParticle, bool> &primaryMCParticleMap)  
+{
+    for (const simb::MCParticle mcParticle : mcParticleVector)
```
Make `mcParticle` a const reference?

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraInput.cxx:
```diff
@@ -469,6 +469,10 @@ void LArPandoraInput::CreatePandoraMCParticles(const Settings &settings, const M
+    // Find Primary Generator Particles
+    std::map<const simb::MCParticle, bool> primaryGeneratorMCParticleMap;
+    LArPandoraInput::FindPrimaryParticles(generatorMCParticleVector, primaryGeneratorMCParticleMap);
+
```
Unnecessary loops for any use cases e.g. uboone, where no generator module label is provided, but I think should otherwise work in case where none of this information is requested. Always good to treble check, though!

PandoraPFA commented on larpandora/LArPandoraInterface/LArPandoraInput.h:
```diff
@@ -89,7 +89,25 @@ class LArPandoraInput
+     *
+     *  @param mcParticleVector vector of all MCParticles to consider
+     *  @param primaryMCParticleMap map containing primary MCParticles and bool indicating whether particle has 
+     *         been accounted for
```
I wouldn't split this over two lines, nor the one on line 107-108

PandoraPFA commented on ups/product_deps:
```diff
@@ -7,7 +7,7 @@ fcldir  product_dir job
 #
 product         version
-larreco         v06_49_00
+larreco         v06_50_00
```
Should remove from pull request.

StevenGreen1 commented:
Hi John,

I've tested these changes on a larsoft v06_63_00 build and everything seems to work.  There was a bug from a missing ! in an if function, but I've corrected that and rebased the history so that this is just a single commit (that doesn't change anything in the ups folder).

Now when testing if you specify the generator module label in the fcl file and, assuming it exists, it access the triggered beam particle information and passes it to Pandora (checked by examining MCParticle nuance code in Pandora xml file).  If this generator label is left blank the information isn't examined and the empty vector/map is propagated, but changes nothing.  

In the future, the next time we make any changes to dunetpc we should also add the default generator module label in the StandardPandora module i.e.

dune_pandora:
{
    module_type:                                                    "StandardPandora"
    GeantModuleLabel:                                               "largeant"
+    GeneratorModuleLabel:                                           ""
...
}

Although this isn't essential at the moment as the c++ sets this default value anyway and this is to just remind people that it's a valid configurable fcl parameter.  Also this is a super minimal change anyway.

Cheers

Steve

PandoraPFA commented:
Now need to cherry-pick commit into a feature branch, push to Redmine as feature/TriggeredMCInfo and mail larsoft-team requesting merge.
